### ISSUE TITLE
Use pubspec.lock instead of hardcoding all the dependency versions inside pubspec.yaml

### ISSUE THREAD

**Issue by isoos** [flutter/flutter#18407]
The specific examples that people noted so far:
- https://github.com/flutter/flutter/blob/master/packages/flutter/pubspec.yaml
- https://github.com/flutter/flutter/blob/master/packages/flutter_test/pubspec.yaml

**Comment by matanlurey** [comment]
@isoos I think the Flutter team did this for a good reason (or bugs with the `.lock` file). You might want to specifically also call out how this is negatively effecting you (i.e. inability to use packages) in this issue.

**Comment by isoos** [comment]
From the pub site's point of view: The transient dependencies that are put in as if they were direct dependencies: they skew the true dependency graph and the popularity metrics, as we treat direct and transient dependencies differently. In the end it causes overpromoting small utility packages that might be used cursorily as part of a utility library's helper method, and at Flutter's level of popularity, this visibly affects the scores.

**Comment by zoechi** [comment]
@Hixie

**Comment by Hixie** [comment]
Lock files in packages have no effect, so this wouldn't work.

**Comment by Hixie** [comment]
(Our goal is to eventually reduce all these dependencies to the bare minimum. Currently the biggest offenders here are `test` and `analyzer` both of which depend on an absurd number of subsequent packages, few of which are actually relevant to Flutter. Ideally those packages would get split into smaller packages that are composed together, and Flutter would only depend on the core parts we need.)

**Comment by isoos** [comment]
> Lock files in packages have no effect, so this wouldn't work.

Could you please expand on this? I'm not sure what part of it wouldn't work.

**Comment by Hixie** [comment]
Lock files only matter for applications. The Flutter framework is a package. If we created a lock file for the Flutter framework, it wouldn't have any impact on what versions of packages would get used by applications. (The goal of the pinning is to make sure that only the versions of packages we have tested get used, since other versions will likely break -- we used to have breaking changes all the time back in the day when we didn't pin everything.)
